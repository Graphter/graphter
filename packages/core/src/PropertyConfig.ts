import { PropertyValidator } from "./PropertyValidator";

export interface PropertyConfig {
  id: string
  name?: string
  description?: string
  type: string
  listItemType?: string
  default?: any | Function
  validation?: PropertyValidator | Array<PropertyValidator>
}