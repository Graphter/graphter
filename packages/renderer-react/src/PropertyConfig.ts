import {ValidationConfig} from "./ValidationConfig";

export interface PropertyConfig {
  id: string,
  name?: string,
  description?: string,
  type: string,
  validations?: Array<ValidationConfig>,
  default?: any | Function
}