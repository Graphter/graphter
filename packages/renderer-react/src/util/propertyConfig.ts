import {PropertyConfig} from "../PropertyConfig";

export function getProp(id: string, properties: Array<PropertyConfig>): PropertyConfig {
  const matchingProp = queryProp(id, properties);
  if(typeof matchingProp === 'undefined') throw new Error(`Couldn't find prop matching '${id}' from layout config. Check the property exists.`);
  return matchingProp;
}

export function queryProp(id: string, properties: Array<PropertyConfig>): PropertyConfig | undefined{
  return properties.find(prop => prop.id === id);
}