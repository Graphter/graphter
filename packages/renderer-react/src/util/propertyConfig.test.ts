import {getProp, queryProp} from "./propertyConfig";
import {PropertyConfig} from "../PropertyConfig";

describe(`propertyConfig utils`, () => {
  let props: Array<PropertyConfig>;
  beforeEach(() => {
    props = [
      {
        id: 'id',
        name: 'ID',
        type: 'string'
      },
      {
        id: 'name',
        name: 'Name',
        type: 'string'
      }
    ];
  });
  describe(`getProp()`, () => {

    it(`should return the value when found`, () => {
      const result = getProp('name', props);
      expect(result).toEqual({
        id: 'name',
        name: 'Name',
        type: 'string'
      });
    });
    it(`should error when the prop can't be found`, () => {
      expect(() => {
        getProp('some-other-prop', props);
      }).toThrowErrorMatchingSnapshot();
    });
  });
  describe(`queryProp()`, () => {
    it(`should return the value when found`, () => {
      const result = queryProp('name', props);
      expect(result).toEqual({
        id: 'name',
        name: 'Name',
        type: 'string'
      });
    });
    it(`should return undefined when the prop can't be found`, () => {
      const result = queryProp('some-other-prop', props);
      expect(result).toBeUndefined();
    });
  });
});


export {};