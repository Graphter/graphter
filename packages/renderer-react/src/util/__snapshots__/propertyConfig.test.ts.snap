// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`propertyConfig utils getProp() should error when the prop can't be found 1`] = `"Couldn't find prop matching 'some-other-prop' from layout config. Check the property exists."`;
